prow/spyglass/podlogartifact.go

```diff
@@ -0,0 +1,167 @@
+	}
+	return &PodLogArtifact{
+		name:      jobName,
+		buildID:   buildID,
+		podName:   podName,
+		sizeLimit: sizeLimit,
+		ja:        ja,
+	}, nil
+}
+
+// CanonicalLink returns a link to where pod logs are streamed
+func (a *PodLogArtifact) CanonicalLink() string {
+	q := url.Values{
+		"job": []string{a.name},
+		"id":  []string{a.buildID},
+	}
+	u := url.URL{
+		Path:     "/log",
+		RawQuery: q.Encode(),
+	}
+	return u.String()
+}
+
+// JobPath gets the path within the job for the pod log. Always returns build-log.txt.
+// This is because the pod log becomes the build log after the job artifact uploads
+// are complete, which should be used instead of the pod log.
+func (a *PodLogArtifact) JobPath() string {
+	return "build-log.txt"
+}
+
+// ReadAt implements reading a range of bytes from the pod logs endpoint
+func (a *PodLogArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	logs := a.fetchLogs()
+	return bytes.NewReader(logs).ReadAt(p, off)
+}
+
+// ReadAll reads all available pod logs, failing if they are too large
+func (a *PodLogArtifact) ReadAll() ([]byte, error) {
+	size, err := a.Size()
+	if err != nil {
+		return []byte{}, nil
+	}
+	if size > a.sizeLimit {
+		return []byte{}, viewers.ErrFileTooLarge
+	}
+	return a.fetchLogs(), nil
+}
+
+// ReadAtMost reads at most n bytes
+func (a *PodLogArtifact) ReadAtMost(n int64) ([]byte, error) {
+	logs := a.fetchLogs()
+	reader := bytes.NewReader(logs)
+	var byteCount int64
+	var p []byte
+	for byteCount <= n {
+		b, err := reader.ReadByte()
+		if err != nil {
+			if err == io.EOF {
+				return p, nil
+			}
+			return p, err
+		}
+		p = append(p, b)
+		byteCount++
+	}
+	return p, nil
+}
+
+// ReadTail reads the last n bytes of the pod log
+func (a *PodLogArtifact) ReadTail(n int64) ([]byte, error) {
+	logs, err := a.ja.GetJobLogTail(a.name, a.buildID, n)
+	if err != nil {
+		logrus.WithField("artifactName", a.name).WithError(err).Error("Error getting pod logs tail")
+	}
+	off := int64(len(logs)) - n - 1
+	p := []byte{}
+	_, err = bytes.NewReader(logs).ReadAt(p, off)
+	if err != nil {
+		logrus.WithError(err).WithField("artifactName", a.name).Error("Failed to read pod logs.")
+	}
+	return p, err
+}
+
+// Size gets the size of the pod log. Note: this function makes the same network call as reading the entire file.
+func (a *PodLogArtifact) Size() (int64, error) {
+	var logs []byte
+	var err error
+	logs, err = a.ja.GetJobLog(a.name, a.buildID)
+	if err != nil {
+		return 0, err
+	}
+	return int64(len(logs)), nil
+
+}
+
+// UseContext is unsupported for pod logs
+func (a *PodLogArtifact) UseContext(ctx context.Context) error {
+	return viewers.ErrContextUnsupported
+}
+
+// fetchLogs is a wrapper method for handling errors from fetching pod logs
+func (a *PodLogArtifact) fetchLogs() []byte {
+	var logs []byte
+	var err error
+	logs, err = a.ja.GetJobLog(a.name, a.buildID)
+	if err != nil {
+		logrus.WithField("artifactName", a.name).WithError(err).Error("Error getting pod logs")
```

Comment:
Don't swallow the error please. I am generally unhappy with this pattern of conflating empty bytes with a failed response